hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBStoreCodecBufferIterator.java

```diff
@@ -84,38 +92,77 @@ CodecBuffer getFromDb() {
     }
   }
 
-  private final Buffer keyBuffer;
-  private final Buffer valueBuffer;
+  private final Stack<RawKeyValue<Buffer>> availableBufferStack;
```

Comment:
Stack is not the best data structure in this case.  We may use a  [BlockingQueue](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/BlockingQueue.html).

It is even better to use  `ThreadLoacl<Buffer>` as mentioned my previous comment.

Reply:
I initially used a queue but I realized we may not have to allocate all buffers all the time. LIFO is better than FIFO in such cases. My buffer allocation would be minimal.

Reply:
Just because we have initialized a Buffer object it doesn't mean we have allocated anything yet. It will still do a lazy allocation. Use of stack will minimize if there are not many concurrent threads in lock code block than the number of max buffers the allocation that is why it is better to use a stack. Moreover stack.pop() is a very cheap operation. There would be only one thread waiting for a non empty stack which should be fine. There should be no contention as well in this case ideally. We need the synchronized block only for the published subscriber model, optmize on the wait() & notify()

Reply:
Instead of having a custom blocking stack implementation I have changed it to using BlockingDeque instead which provides these functionalities. As part of this change I had to handle race conditions with the inuseBufferSet values. Since we need to handle 2 data structures in a consistently. 
Please take a look at this and check which is lesser complex we can go ahead with whatever approach we all align with.

Reply:
> ... LIFO is better than FIFO in such cases. ...

Why LIFO is better than FIFO? 

BTW, for LIFO, we have https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/BlockingDeque.html

> ... My buffer allocation would be minimal with a stack

For any data structures, the allocation can be minimal since we can have lazy allocation -- allocate only when it is used.

> ... We need the synchronized block only for the published subscriber model, optmize on the wait() & notify()

It also makes the code complicated.  Prone to a deadlock.